Build error on node 12

I can provide more details later - on a time crunch at the moment. But just for people trying to use the lib and getting build errors while installing the package wit NPM - just roll back to using node 11 for now.

Will try to capture logs for debugging later.

I experienced the same issue with Node 13 and believe that this error is indeed related to #129. I can provide a pull request which works with Nodejs v13 (all tests are passing on Linux with a Canon EOS 700D), but I do not have time to test backward compatibility. 

Caveats: 
* I did not fix the compiler warnings
* I have no experience with V8 or Nan

```
$ node --version
v13.12.0

$ node-gyp --version
v6.1.0

$ npm --version
6.14.4

$ uname -a
Linux test 5.4.30-1-MANJARO #1 SMP PREEMPT Thu Apr 2 17:31:45 UTC 2020 x86_64 GNU/Linux

$ clang --version
clang version 9.0.1 
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin

$ pkg-config --modversion libgphoto2
2.5.24
```

Here's the build output before fixing the code:

```
In file included from ../src/././binding.h:12,
                 from ../src/./camera.h:10,
                 from ../src/camera.cc:8:
../node_modules/nan/nan.h: In function ‘void Nan::AsyncQueueWorker(Nan::AsyncWorker*)’:
../node_modules/nan/nan.h:2298:62: warning: cast between incompatible function types from ‘void (*)(uv_work_t*)’ {aka ‘void (*)(uv_work_s*)’} to ‘uv_after_work_cb’ {aka ‘void (*)(uv_work_s*, int)’} [-Wcast-function-type]
 2298 |     , reinterpret_cast<uv_after_work_cb>(AsyncExecuteComplete)
      |                                                              ^
../src/camera.cc: In static member function ‘static void GPCamera::Initialize(Nan::ADDON_REGISTER_FUNCTION_ARGS_TYPE)’:
../src/camera.cc:43:38: error: no matching function for call to ‘v8::FunctionTemplate::GetFunction()’
   43 |   constructor.Reset(tpl->GetFunction());
      |                                      ^
In file included from ../src/././binding.h:11,
                 from ../src/./camera.h:10,
                 from ../src/camera.cc:8:
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:6283:46: note: candidate: ‘v8::MaybeLocal<v8::Function> v8::FunctionTemplate::GetFunction(v8::Local<v8::Context>)’
 6283 |   V8_WARN_UNUSED_RESULT MaybeLocal<Function> GetFunction(
      |                                              ^~~~~~~~~~~
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:6283:46: note:   candidate expects 1 argument, 0 provided
../src/camera.cc: In static member function ‘static Nan::NAN_METHOD_RETURN_TYPE GPCamera::DownloadPicture(Nan::NAN_METHOD_ARGS_TYPE)’:
../src/camera.cc:173:85: error: no matching function for call to ‘v8::Object::Get(v8::Local<v8::String>)’
  173 |   v8::Local<v8::Value> source = options->Get(Nan::New("cameraPath").ToLocalChecked());
      |                                                                                     ^
In file included from ../src/././binding.h:11,
                 from ../src/./camera.h:10,
                 from ../src/camera.cc:8:
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:3594:43: note: candidate: ‘v8::MaybeLocal<v8::Value> v8::Object::Get(v8::Local<v8::Context>, v8::Local<v8::Value>)’
 3594 |   V8_WARN_UNUSED_RESULT MaybeLocal<Value> Get(Local<Context> context,
      |                                           ^~~
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:3594:43: note:   candidate expects 2 arguments, 1 provided
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:3597:43: note: candidate: ‘v8::MaybeLocal<v8::Value> v8::Object::Get(v8::Local<v8::Context>, uint32_t)’
 3597 |   V8_WARN_UNUSED_RESULT MaybeLocal<Value> Get(Local<Context> context,
      |                                           ^~~
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:3597:43: note:   candidate expects 2 arguments, 1 provided
../src/camera.cc:174:85: error: no matching function for call to ‘v8::Object::Get(v8::Local<v8::String>)’
  174 |   v8::Local<v8::Value> target = options->Get(Nan::New("targetPath").ToLocalChecked());
      |                                                                                     ^
In file included from ../src/././binding.h:11,
                 from ../src/./camera.h:10,
                 from ../src/camera.cc:8:
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:3594:43: note: candidate: ‘v8::MaybeLocal<v8::Value> v8::Object::Get(v8::Local<v8::Context>, v8::Local<v8::Value>)’
 3594 |   V8_WARN_UNUSED_RESULT MaybeLocal<Value> Get(Local<Context> context,
      |                                           ^~~
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:3594:43: note:   candidate expects 2 arguments, 1 provided
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:3597:43: note: candidate: ‘v8::MaybeLocal<v8::Value> v8::Object::Get(v8::Local<v8::Context>, uint32_t)’
 3597 |   V8_WARN_UNUSED_RESULT MaybeLocal<Value> Get(Local<Context> context,
      |                                           ^~~
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:3597:43: note:   candidate expects 2 arguments, 1 provided
../src/camera.cc: In static member function ‘static Nan::NAN_METHOD_RETURN_TYPE GPCamera::New(Nan::NAN_METHOD_ARGS_TYPE)’:
../src/camera.cc:360:105: error: no matching function for call to ‘v8::Object::Set(v8::Local<v8::String>, v8::Local<v8::String>)’
  360 |   info.This()->Set(Nan::New("model").ToLocalChecked(), Nan::New(camera->model_.c_str()).ToLocalChecked());
      |                                                                                                         ^
In file included from ../src/././binding.h:11,
                 from ../src/./camera.h:10,
                 from ../src/camera.cc:8:
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:3547:37: note: candidate: ‘v8::Maybe<bool> v8::Object::Set(v8::Local<v8::Context>, v8::Local<v8::Value>, v8::Local<v8::Value>)’
 3547 |   V8_WARN_UNUSED_RESULT Maybe<bool> Set(Local<Context> context,
      |                                     ^~~
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:3547:37: note:   candidate expects 3 arguments, 2 provided
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:3550:37: note: candidate: ‘v8::Maybe<bool> v8::Object::Set(v8::Local<v8::Context>, uint32_t, v8::Local<v8::Value>)’
 3550 |   V8_WARN_UNUSED_RESULT Maybe<bool> Set(Local<Context> context, uint32_t index,
      |                                     ^~~
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:3550:37: note:   candidate expects 3 arguments, 2 provided
../src/camera.cc:361:103: error: no matching function for call to ‘v8::Object::Set(v8::Local<v8::String>, v8::Local<v8::String>)’
  361 |   info.This()->Set(Nan::New("port").ToLocalChecked(), Nan::New(camera->port_.c_str()).ToLocalChecked());
      |                                                                                                       ^
In file included from ../src/././binding.h:11,
                 from ../src/./camera.h:10,
                 from ../src/camera.cc:8:
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:3547:37: note: candidate: ‘v8::Maybe<bool> v8::Object::Set(v8::Local<v8::Context>, v8::Local<v8::Value>, v8::Local<v8::Value>)’
 3547 |   V8_WARN_UNUSED_RESULT Maybe<bool> Set(Local<Context> context,
      |                                     ^~~
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:3547:37: note:   candidate expects 3 arguments, 2 provided
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:3550:37: note: candidate: ‘v8::Maybe<bool> v8::Object::Set(v8::Local<v8::Context>, uint32_t, v8::Local<v8::Value>)’
 3550 |   V8_WARN_UNUSED_RESULT Maybe<bool> Set(Local<Context> context, uint32_t index,
      |                                     ^~~
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:3550:37: note:   candidate expects 3 arguments, 2 provided
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h: In instantiation of ‘void v8::PersistentBase<T>::SetWeak(P*, typename v8::WeakCallbackInfo<P>::Callback, v8::WeakCallbackType) [with P = node::ObjectWrap; T = v8::Object; typename v8::WeakCallbackInfo<P>::Callback = void (*)(const v8::WeakCallbackInfo<node::ObjectWrap>&)]’:
/home/test/.cache/node-gyp/13.12.0/include/node/node_object_wrap.h:85:78:   required from here
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:10400:16: warning: cast between incompatible function types from ‘v8::WeakCallbackInfo<node::ObjectWrap>::Callback’ {aka ‘void (*)(const v8::WeakCallbackInfo<node::ObjectWrap>&)’} to ‘Callback’ {aka ‘void (*)(const v8::WeakCallbackInfo<void>&)’} [-Wcast-function-type]
10400 |                reinterpret_cast<Callback>(callback), type);
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h: In instantiation of ‘void v8::PersistentBase<T>::SetWeak(P*, typename v8::WeakCallbackInfo<P>::Callback, v8::WeakCallbackType) [with P = Nan::ObjectWrap; T = v8::Object; typename v8::WeakCallbackInfo<P>::Callback = void (*)(const v8::WeakCallbackInfo<Nan::ObjectWrap>&)]’:
../node_modules/nan/nan_object_wrap.h:65:61:   required from here
/home/test/.cache/node-gyp/13.12.0/include/node/v8.h:10400:16: warning: cast between incompatible function types from ‘v8::WeakCallbackInfo<Nan::ObjectWrap>::Callback’ {aka ‘void (*)(const v8::WeakCallbackInfo<Nan::ObjectWrap>&)’} to ‘Callback’ {aka ‘void (*)(const v8::WeakCallbackInfo<void>&)’} [-Wcast-function-type]
make: *** [gphoto2.target.mk:113: Release/obj.target/gphoto2/src/camera.o] Error 1

```

I'm trying to build `node-gphoto2` against Node v12, also having some issue:

```
error /Users/philipp/git/photo-booth-new/node_modules/gphoto2: Command failed.
Exit code: 1
Command: node-gyp rebuild
Arguments:
Directory: /Users/philipp/git/photo-booth-new/node_modules/gphoto2
Output:
gyp info it worked if it ends with ok
gyp info using node-gyp@3.8.0
gyp info using node@12.1.0 | darwin | x64
gyp info spawn /usr/local/bin/python2
gyp info spawn args [
gyp info spawn args   '/usr/local/Cellar/node/12.1.0/libexec/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/philipp/git/photo-booth-new/node_modules/gphoto2/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/usr/local/Cellar/node/12.1.0/libexec/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/philipp/.node-gyp/12.1.0/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/Users/philipp/.node-gyp/12.1.0',
gyp info spawn args   '-Dnode_gyp_dir=/usr/local/Cellar/node/12.1.0/libexec/lib/node_modules/npm/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=/Users/philipp/.node-gyp/12.1.0/<(target_arch)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=/Users/philipp/git/photo-booth-new/node_modules/gphoto2',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.'
gyp info spawn args ]
gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
  CXX(target) Release/obj.target/gphoto2/src/autodetect.o
  CXX(target) Release/obj.target/gphoto2/src/binding.o
  CXX(target) Release/obj.target/gphoto2/src/camera.o
../src/camera.cc:43:38: error: too few arguments to function call, single argument 'context' was not specified
  constructor.Reset(tpl->GetFunction());
                    ~~~~~~~~~~~~~~~~ ^
/Users/philipp/.node-gyp/12.1.0/include/node/v8.h:5947:3: note: 'GetFunction' declared here
  V8_WARN_UNUSED_RESULT MaybeLocal<Function> GetFunction(
  ^
/Users/philipp/.node-gyp/12.1.0/include/node/v8config.h:347:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/camera.cc:173:42: warning: 'Get' is deprecated: Use maybe version [-Wdeprecated-declarations]
  v8::Local<v8::Value> source = options->Get(Nan::New("cameraPath").ToLocalChecked());
                                         ^
/Users/philipp/.node-gyp/12.1.0/include/node/v8.h:3412:3: note: 'Get' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version", Local<Value> Get(Local<Value> key));
  ^
/Users/philipp/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/camera.cc:174:42: warning: 'Get' is deprecated: Use maybe version [-Wdeprecated-declarations]
  v8::Local<v8::Value> target = options->Get(Nan::New("targetPath").ToLocalChecked());
                                         ^
/Users/philipp/.node-gyp/12.1.0/include/node/v8.h:3412:3: note: 'Get' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version", Local<Value> Get(Local<Value> key));
  ^
/Users/philipp/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/camera.cc:360:16: warning: 'Set' is deprecated: Use maybe version [-Wdeprecated-declarations]
  info.This()->Set(Nan::New("model").ToLocalChecked(), Nan::New(camera->model_.c_str()).ToLocalChecked());
               ^
/Users/philipp/.node-gyp/12.1.0/include/node/v8.h:3358:3: note: 'Set' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version",
  ^
/Users/philipp/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/camera.cc:361:16: warning: 'Set' is deprecated: Use maybe version [-Wdeprecated-declarations]
  info.This()->Set(Nan::New("port").ToLocalChecked(), Nan::New(camera->port_.c_str()).ToLocalChecked());
               ^
/Users/philipp/.node-gyp/12.1.0/include/node/v8.h:3358:3: note: 'Set' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version",
  ^
/Users/philipp/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
4 warnings and 1 error generated.
make: *** [Release/obj.target/gphoto2/src/camera.o] Error 1
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/Cellar/node/12.1.0/libexec/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:262:23)
gyp ERR! stack     at ChildProcess.emit (events.js:196:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:256:12)
gyp ERR! System Darwin 19.4.0
gyp ERR! command "/usr/local/Cellar/node/12.1.0/bin/node" "/usr/local/Cellar/node/12.1.0/libexec/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/philipp/git/photo-booth-new/node_modules/gphoto2
``` 

The issue seems to be `GetFunction()`, which needs `context` as argument.

--- a/.gitignore
+++ b/.gitignore
@@ -8,3 +8,4 @@ npm-debug.log
 .idea
 yarn-error.log
 .npmrc
+.vscode/

--- a/package.json
+++ b/package.json
@@ -19,7 +19,7 @@
   ],
   "main": "index.js",
   "dependencies": {
-    "nan": "^2.11.1"
+    "nan": "^2.14.0"
   },
   "devDependencies": {
     "async": "^2.6.1",

--- a/src/camera.cc
+++ b/src/camera.cc
@@ -40,7 +40,7 @@ NAN_MODULE_INIT(GPCamera::Initialize) {
   Nan::SetPrototypeMethod(tpl, "takePicture", TakePicture);
   Nan::SetPrototypeMethod(tpl, "downloadPicture", DownloadPicture);
 
-  constructor.Reset(tpl->GetFunction());
+  constructor.Reset(Nan::GetFunction(tpl).ToLocalChecked());
 
   Nan::Set(target, Nan::New("Camera").ToLocalChecked(), Nan::GetFunction(tpl).ToLocalChecked());
 }
@@ -170,8 +170,8 @@ NAN_METHOD(GPCamera::DownloadPicture) {
   gp_context_ref(picture_req->context);
   picture_req->download = true;
 
-  v8::Local<v8::Value> source = options->Get(Nan::New("cameraPath").ToLocalChecked());
-  v8::Local<v8::Value> target = options->Get(Nan::New("targetPath").ToLocalChecked());
+  v8::Local<v8::Value> source = Nan::Get(options, Nan::New("cameraPath").ToLocalChecked()).ToLocalChecked();
+  v8::Local<v8::Value> target = Nan::Get(options, Nan::New("targetPath").ToLocalChecked()).ToLocalChecked();
   if (target->IsString()) {
     picture_req->target_path = std::string(*Nan::Utf8String(target));
   }
@@ -357,8 +357,8 @@ NAN_METHOD(GPCamera::New) {
   GPCamera *camera = new GPCamera(js_gphoto, (std::string) *model_, (std::string) *port_);
 
   camera->Wrap(info.This());
-  info.This()->Set(Nan::New("model").ToLocalChecked(), Nan::New(camera->model_.c_str()).ToLocalChecked());
-  info.This()->Set(Nan::New("port").ToLocalChecked(), Nan::New(camera->port_.c_str()).ToLocalChecked());
+  Nan::Set(info.This(), Nan::New("model").ToLocalChecked(), Nan::New(camera->model_.c_str()).ToLocalChecked());
+  Nan::Set(info.This(), Nan::New("port").ToLocalChecked(), Nan::New(camera->port_.c_str()).ToLocalChecked());
   info.GetReturnValue().Set(info.This());
 }
 

--- a/src/camera_helpers.cc
+++ b/src/camera_helpers.cc
@@ -62,16 +62,16 @@ v8::Local<v8::Value> GPCamera::getWidgetValue(GPContext *context, CameraWidget *
   ret = gp_widget_get_label(widget, &label);
   if (ret != GP_OK) return Nan::Undefined();
 
-  value->Set(Nan::New("label").ToLocalChecked(), Nan::New(label).ToLocalChecked());
-  value->Set(Nan::New("type").ToLocalChecked(), Nan::Undefined());
+  Nan::Set(value, Nan::New("label").ToLocalChecked(), Nan::New(label).ToLocalChecked());
+  Nan::Set(value, Nan::New("type").ToLocalChecked(), Nan::Undefined());
 
   switch (type) {
   case GP_WIDGET_TEXT: { /* char * */
     char *txt;
     ret = gp_widget_get_value(widget, &txt);
     if (ret == GP_OK) {
-      value->Set(Nan::New("type").ToLocalChecked(), Nan::New("string").ToLocalChecked());
-      value->Set(Nan::New("value").ToLocalChecked(), Nan::New(txt).ToLocalChecked());
+      Nan::Set(value, Nan::New("type").ToLocalChecked(), Nan::New("string").ToLocalChecked());
+      Nan::Set(value, Nan::New("value").ToLocalChecked(), Nan::New(txt).ToLocalChecked());
     } else {
       gp_context_error(context,
                        "Failed to retrieve value of text widget %s.",
@@ -85,11 +85,11 @@ v8::Local<v8::Value> GPCamera::getWidgetValue(GPContext *context, CameraWidget *
     ret = gp_widget_get_range(widget, &min, &max, &step);
     if (ret == GP_OK) {
       ret = gp_widget_get_value(widget, &f);
-      value->Set(Nan::New("type").ToLocalChecked(), Nan::New("range").ToLocalChecked());
-      value->Set(Nan::New("value").ToLocalChecked(), Nan::New(f));
-      value->Set(Nan::New("max").ToLocalChecked(), Nan::New(max));
-      value->Set(Nan::New("min").ToLocalChecked(), Nan::New(min));
-      value->Set(Nan::New("step").ToLocalChecked(), Nan::New(step));
+      Nan::Set(value, Nan::New("type").ToLocalChecked(), Nan::New("range").ToLocalChecked());
+      Nan::Set(value, Nan::New("value").ToLocalChecked(), Nan::New(f));
+      Nan::Set(value, Nan::New("max").ToLocalChecked(), Nan::New(max));
+      Nan::Set(value, Nan::New("min").ToLocalChecked(), Nan::New(min));
+      Nan::Set(value, Nan::New("step").ToLocalChecked(), Nan::New(step));
     } else {
       gp_context_error(context,
                        "Failed to retrieve values of range widget %s.",
@@ -101,8 +101,8 @@ v8::Local<v8::Value> GPCamera::getWidgetValue(GPContext *context, CameraWidget *
     int t;
     ret = gp_widget_get_value(widget, &t);
     if (ret == GP_OK) {
-      value->Set(Nan::New("type").ToLocalChecked(), Nan::New("toggle").ToLocalChecked());
-      value->Set(Nan::New("value").ToLocalChecked(), Nan::New(t));
+      Nan::Set(value, Nan::New("type").ToLocalChecked(), Nan::New("toggle").ToLocalChecked());
+      Nan::Set(value, Nan::New("value").ToLocalChecked(), Nan::New(t));
     } else {
       gp_context_error(context,
                        "Failed to retrieve values of toggle widget %s.",
@@ -119,8 +119,8 @@ v8::Local<v8::Value> GPCamera::getWidgetValue(GPContext *context, CameraWidget *
                        label);
       break;
     }
-    value->Set(Nan::New("type").ToLocalChecked(), Nan::New("date").ToLocalChecked());
-    value->Set(Nan::New("value").ToLocalChecked(), Nan::New<v8::Date>(t * 1000.0).ToLocalChecked());
+    Nan::Set(value, Nan::New("type").ToLocalChecked(), Nan::New("date").ToLocalChecked());
+    Nan::Set(value, Nan::New("value").ToLocalChecked(), Nan::New<v8::Date>(t * 1000.0).ToLocalChecked());
     break;
   }
   case GP_WIDGET_MENU:
@@ -142,25 +142,25 @@ v8::Local<v8::Value> GPCamera::getWidgetValue(GPContext *context, CameraWidget *
 
       ret = gp_widget_get_choice(widget, i, &choice);
       if (ret != GP_OK) continue;
-      choices->Set(Nan::New(i), Nan::New(choice).ToLocalChecked());
+      Nan::Set(choices, Nan::New(i), Nan::New(choice).ToLocalChecked());
     }
 
-    value->Set(Nan::New("type").ToLocalChecked(), Nan::New("choice").ToLocalChecked());
-    value->Set(Nan::New("value").ToLocalChecked(), Nan::New(current).ToLocalChecked());
-    value->Set(Nan::New("choices").ToLocalChecked(), choices);
+    Nan::Set(value, Nan::New("type").ToLocalChecked(), Nan::New("choice").ToLocalChecked());
+    Nan::Set(value, Nan::New("value").ToLocalChecked(), Nan::New(current).ToLocalChecked());
+    Nan::Set(value, Nan::New("choices").ToLocalChecked(), choices);
     break;
   }
   /* ignore: */
   case GP_WIDGET_WINDOW: {
-    value->Set(Nan::New("type").ToLocalChecked(), Nan::New("window").ToLocalChecked());
+    Nan::Set(value, Nan::New("type").ToLocalChecked(), Nan::New("window").ToLocalChecked());
     break;
   }
   case GP_WIDGET_SECTION: {
-    value->Set(Nan::New("type").ToLocalChecked(), Nan::New("section").ToLocalChecked());
+    Nan::Set(value, Nan::New("type").ToLocalChecked(), Nan::New("section").ToLocalChecked());
     break;
   }
   case GP_WIDGET_BUTTON: {
-    value->Set(Nan::New("type").ToLocalChecked(), Nan::New("button").ToLocalChecked());
+    Nan::Set(value, Nan::New("type").ToLocalChecked(), Nan::New("button").ToLocalChecked());
   }
   }
   return scope.Escape(value);

--- a/src/gphoto.cc
+++ b/src/gphoto.cc
@@ -52,7 +52,7 @@ NAN_MODULE_INIT(GPhoto2::Initialize) {
   Nan::SetPrototypeMethod(tpl, "list", List);
   Nan::SetPrototypeMethod(tpl, "setLogLevel", SetLogLevel);
 
-  constructor.Reset(tpl->GetFunction());
+  constructor.Reset(Nan::GetFunction(tpl).ToLocalChecked());
 
   Nan::Set(target, Nan::New("GPhoto2").ToLocalChecked(), Nan::GetFunction(tpl).ToLocalChecked());
 }
@@ -197,8 +197,8 @@ void GPhoto2::Async_ListCb(uv_work_t *req, int status) {
     // call the _javascript_ constructor to create a new Camera object
     v8::Local<v8::Object> js_camera = Nan::NewInstance(Nan::New(GPCamera::constructor), 3, _argv).ToLocalChecked();
 
-    js_camera->Set(Nan::New("_gphoto2_ref_obj_").ToLocalChecked(), Nan::New(list_req->This));
-    result->Set(Nan::New(i), js_camera);
+    Nan::Set(js_camera, Nan::New("_gphoto2_ref_obj_").ToLocalChecked(), Nan::New(list_req->This));
+    Nan::Set(result, Nan::New(i), js_camera);
     if (try_catch.HasCaught()) {
       goto finally;
     }

